File: java/src/main/java/com/example/cote/PGMRS/징검다리.java
package com.example.cote.PGMRS;

import java.util.*;
import java.util.stream.*;

public class 징검다리{


	private static final int TEST_CASES = 10;

	private static final int MAX_DISTANCE =  1000000000;
	private static final int MAX_ROCK_NUMBER = 50000;

	private static final Random rd = new Random(43);
	
	private static int distance;
	private static int r;
	private static int n;
	private static int[] rocks;


	private static void initialize(){
		distance = rd.nextInt(MAX_DISTANCE-1) +1 ;
		r = rd.nextInt(MAX_ROCK_NUMBER-1)+1;
		n = rd.nextInt(r-2)+1;
		rocks = new int[n];

		for(int i=0; i<n; i++)
			rocks[i] = rd.nextInt(distance-2) +1;
	}

	private static  boolean isValid(int d) {

		int removed =0; 
		int last = 0; 
		
		for(int i=0; i<rocks.length; i++) {
			if(rocks[i] -last < d){
				removed++;
				continue; 
			}
			last = rocks[i];
		}
		return removed <= n;

	}

	private static int solve(){
		rocks = Arrays.copyOf(rocks, rocks.length + 1);
		rocks[rocks.length-1] = distance;
		Arrays.sort(rocks);

		int lo = 1; 
		int hi = distance+1;

		while(hi - lo > 1){
			int mid= (hi+lo)/2; 
			if(isValid(mid))
				lo = mid;
			else 
				hi = mid;

		}

		return lo;
	}

	public static void main(String... args){

		for (int it =0; it <TEST_CASES; it++){
			initialize();
			System.out.println(solve());

		}

	}
}

File: go/cote/PGMRS/징검다리.go
package main

import (
	"fmt"
	"math/rand"
	"slices"
)

const (
	TEST_CASE       = 10
	MAX_DISTANCE    = 1000000000
	MAX_ROCK_NUMBER = 50000
)

func isValid(rocks []int, n, distance int) bool {

	last, removed := 0, 0

	for _, rock := range rocks {
		if rock-last < distance {
			removed++
			continue
		}
		last = rock
	}

	return removed <= n
}

func solve(rocks []int, n int) int {

	lo, hi := 0, MAX_DISTANCE

	for {
		if lo+1 == hi {
			break
		}

		mid := int((lo + hi) / 2)

		if isValid(rocks, n, mid) {
			lo = mid
		} else {
			hi = mid
		}
	}

	return lo
}

func main() {

	for it := 0; it < TEST_CASE; it++ {
		r := rand.Intn(MAX_ROCK_NUMBER-1) + 1
		n := rand.Intn(r-1) + 1

		rocks := make([]int, r)

		for i := range rocks {
			rocks[i] = rand.Intn(MAX_DISTANCE)
		}

		slices.Sort(rocks)

		fmt.Println(solve(rocks, n))
	}
}
